clp_ffi_py/query_builder.py

```diff
@@ -0,0 +1,182 @@
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import List, Optional
+
+from clp_ffi_py.ir import Query
+from clp_ffi_py.wildcard_query import WildcardQuery
+
+
+class QueryBuilderException(Exception):
+    """
+    This exception will be raised when illegal parameters are used to build a
+    Query object.
+    """
+
+    pass
+
+
+class QueryBuilder:
+    """
+    This class serves as an interface for conveniently constructing Query
+    objects utilized in CLP IR streaming search. It provides methods for.
+
+    configuring and resetting search parameters, which encompass:
+        - search_time_lower_bound: Start of search time range (inclusive).
+        - search_time_upper_bound: End of search time range (inclusive).
+        - search_time_termination_margin: The margin used to determine the
+          search termination timestamp.
+        - wildcard_queries: A list of wildcard search queries.
+
+    Timestamps are specified using Unix epoch timestamps in milliseconds. By
+    default, the lower and upper bounds for search timestamps encompass the
+    entire valid range of Unix epoch timestamps, while the wildcard queries list
+    is empty.
+
+    For more details about the search query CLP IR stream supports, please read
+    the documents of `clp_ffi_py.ir.Query` and
+    `clp_ffi_py.wildcard_query.WildcardQuery`.
+    """
+
+    def __init__(self) -> None:
+        self._search_time_lower_bound: int = Query.default_search_time_lower_bound()
+        self._search_time_upper_bound: int = Query.default_search_time_upper_bound()
+        self._search_time_termination_margin: int = Query.default_search_time_termination_margin()
+        self._wildcard_queries: List[WildcardQuery] = []
+
+    @property
+    def search_time_lower_bound(self) -> int:
+        return self._search_time_lower_bound
+
+    @property
+    def search_time_upper_bound(self) -> int:
+        return self._search_time_upper_bound
+
+    @property
+    def search_time_termination_margin(self) -> int:
+        return self._search_time_termination_margin
+
+    @property
+    def wildcard_queries(self) -> List[WildcardQuery]:
+        return deepcopy(self._wildcard_queries)
```

Comment:
You should document that you're returning a deep copy since that's not going to be obvious to all users.